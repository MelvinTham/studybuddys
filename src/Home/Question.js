import React, { PureComponent } from 'react';
import {
    Typography,
    Paper,
    TextField,
    Button,
} from '@material-ui/core';
import './CoursePage.css';
import axios from 'axios';
import * as firebase from 'firebase';
import Navbar from "./Navbar";
// import Calendar from "./Calendar2";
// import Calendar from 'rc-calendar';
// import LuxonUtils from '@date-io/luxon';
// import { MuiPickersUtilsProvider } from 'material-ui-pickers';

class Question extends PureComponent {
    constructor(props) {
        super(props);
        this.state = {
            name:"",
            course: '',
            questID: "",
            question: "",
            createdBy: "",
            replyText: "",
            replies: [],
            replier: [],
        }
        this.checkLoggedIn = this.checkLoggedIn.bind(this);
        this.handleChange = this.handleChange.bind(this);
        this.logout = this.logout.bind(this);
        this.getUserName = this.getUserName.bind(this);
        this.getQuestion = this.getQuestion.bind(this);
        this.submitAnswer = this.submitAnswer.bind(this);
    }
    componentDidMount() {
        this.checkLoggedIn();
        const { courseName } = this.props.match.params;
        const { questionID } = this.props.match.params;
        fetch(`/course/${courseName}?/${questionID}?`)
            .then(
                this.setState({
                    course : courseName, 
                    questID: questionID
                })
            )
            .then (
                this.getQuestion,
                this.getReplies(questionID)
            )
        this.getUserName();
        
    }

    checkLoggedIn(){
        var prop = this.props;
        firebase.auth().onAuthStateChanged(function(user) {
            if (!user) {
            prop.history.push('/');
            }
        });
  }

    handleChange = name => event => {
        this.setState({
            [name]: event.target.value,
        });
    };
    
    logout(e){
        e.preventDefault();
        firebase.auth().signOut();
        this.props.history.push('/');
    }
    getUserName(e){
        var page = this;
        firebase.auth().onAuthStateChanged(function(user) {
        if (user) {
            var info = {
                id: user.uid
            }
            axios.post('https://triple-bonito-221722.appspot.com/api/getUsername', info)
            .then(response => {
                page.setState({name: response.data.name})
            })
        }
        });
    }

    submitAnswer() {
        var reply = this.state.replyText;
        var questID = this.state.questID;
        firebase.auth().onAuthStateChanged(function(user) {
            if (user) {
                var info = {
                    id: user.uid,
                    replyText: reply,
                    questionID: questID,
                }
                axios.post('https://triple-bonito-221722.appspot.com/api/submitAnswer', info)
            }
        });
        this.getReplies(questID);
    }
    
    getReplies(ID) {
        var info = {
            questionID: ID,
        }
        axios.post('https://triple-bonito-221722.appspot.com/api/getReplies', info)
        .then( response => {
            this.setState({
                replies: response.data.replies,
                replier: response.data.names
            })
        })
    }
    getQuestion() {
        var questID = {
            courseName: this.state.course,
            questionID: this.state.questID
        };
        axios.post('https://triple-bonito-221722.appspot.com/api/getSingleQuestion', questID)
            .then(response => {
                console.log(response.data)
                this.setState({
                    question: response.data.question,
                    createdBy: response.data.name
                })
            })
            
    }
    

    render() {
        return (
            <div data-aos ="fade-in" data-aos-easing="linear" data-aos-duration="800" style = {{display: "flex", flexDirection: "column"}}>
                <div>
                    <div style = {{float: "right", display: "inline-block"}}>
                        <span>{this.state.name}</span>
                        <Button onClick={this.logout}>Logout</Button>
                    </div>
                    <Navbar/>
                </div>
                    {/* <Button 
                        className="Calendar"
                        type="submit"
                        onClick={this.openCalendar}>
                        Meet Up
                    </Button>
                {
                    this.state.calendarIsOpen
                    ?
                    <MuiPickersUtilsProvider 
                        utils={LuxonUtils}>
                        <Calendar />
                    </MuiPickersUtilsProvider>
                    : null
                } */}

                <Typography variant = "h1" style = {{margin: "16px auto"}}><strong>{this.state.question}</strong></Typography>
                <Typography variant = "h6" style = {{margin: "0px auto"}}><em>Created By: {this.state.createdBy}</em></Typography>
                <div className = "flexCenter">
                    {this.state.replies.map((data, key) => {
                        return (
                            <Paper className = "flexCenter" style = {{margin: "10px auto", width: "65%", height: "10%"}}>
                                <Typography gutterBottom = {true} variant = "subtitle">
                                    <em>{data}</em>
                                </Typography>
                                <Typography variant = "subtitle">
                                    <em>Answered By: {this.state.replier[key]}</em>
                                </Typography>
                            </Paper>
                        )
                    })}
                    <TextField 
                        variant = "filled" 
                        multiline = {true} 
                        label = "Answer" 
                        style = {{marginTop: "20px", width: "80%"}}
                        onChange= {                                         
                            this.handleChange("replyText")         
                        }
                    >
                    {console.log(this.state.replyText) }
                    </TextField>
                    <Button type = "submit" onClick = {this.submitAnswer} style = {{width: "80%"}}>Submit</Button>
                </div>
            </div>
         )   
        }
        
    }

export default Question;