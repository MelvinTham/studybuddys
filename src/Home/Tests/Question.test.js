import Enzyme, { render, shallow } from 'enzyme';
import enzymeAdapterReact16 from 'enzyme-adapter-react-16';
import React from 'react';
import Question from './../Question';
import { BrowserRouter as Router } from 'react-router-dom';
import toJson from 'enzyme-to-json';

Enzyme.configure({ adapter: new enzymeAdapterReact16() });
const getDefaultProps = () => ({});
describe('Question component', () => {
  it('shallow renders without crashing', () => {
    const {} = getDefaultProps();
    shallow(
        <Router>
            <Question />
        </Router>);
  });
  it('render snapshot', () => {
    const {} = getDefaultProps();
    const wrapper = render(
        <Router>
            <Question />
        </Router>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
