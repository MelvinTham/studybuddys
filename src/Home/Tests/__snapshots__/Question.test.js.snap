// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question component render snapshot 1`] = `
<div
  data-aos="fade-in"
  data-aos-duration="800"
  data-aos-easing="linear"
>
  <div
    class="nav"
  >
    <div>
      <button
        class="MuiButtonBase-root-28 MuiButton-root-2 MuiButton-text-4 MuiButton-flat-7"
        style="float:left"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label-3"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-31 MuiSvgIcon-fontSizeLarge-39"
            focusable="false"
            role="presentation"
            style="color:white"
            viewBox="0 0 24 24"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
            />
          </svg>
        </span>
      </button>
    </div>
    <div
      class="flexRow"
      style="margin-left:auto"
    >
      <span
        style="font-size:22px;margin-right:10px"
      >
        <a
          href="/dashboard/profile"
          style="color:white"
        />
      </span>
      <button
        class="MuiButtonBase-root-28 MuiButton-root-2 MuiButton-text-4 MuiButton-flat-7"
        style="color:white;font-size:17px"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label-3"
        >
          Logout
        </span>
      </button>
    </div>
  </div>
  <br />
  <h2
    class="MuiTypography-root-54 MuiTypography-h2-67"
    style="margin:16px auto"
  >
    <strong />
  </h2>
  <h6
    class="MuiTypography-root-54 MuiTypography-h6-71"
    style="margin:0px auto"
  >
    <em>
      Created By: 
    </em>
  </h6>
  <div
    class="flexCenter"
    style="margin:30px;background-color:#ffffff;border:1px solid black;padding:20px;border-radius:10px;box-shadow:5px 5px 5px 5px #777777;-moz-box-shadow:0 0 10px #777777;-webkit-box-shadow:0 0 10px #777777"
  >
    <div
      class="MuiFormControl-root-89"
      style="width:50%;margin-top:15px"
    >
      <label
        class="MuiFormLabel-root-104 MuiInputLabel-root-93 MuiInputLabel-formControl-98 MuiInputLabel-animated-101 MuiInputLabel-outlined-103"
        data-shrink="false"
      >
        Answer
      </label>
      <div
        class="MuiInputBase-root-124 MuiOutlinedInput-root-111 MuiInputBase-formControl-125 MuiInputBase-multiline-132 MuiOutlinedInput-multiline-117"
      >
        <fieldset
          aria-hidden="true"
          class="MuiPrivateNotchedOutline-root-141 MuiOutlinedInput-notchedOutline-118"
          style="padding-left:8px"
        >
          <legend
            class="MuiPrivateNotchedOutline-legend-142"
            style="width:0.01px"
          >
            <span>
              â€‹
            </span>
          </legend>
        </fieldset>
        <div
          class="MuiPrivateTextarea-root-143"
        >
          <textarea
            aria-hidden="true"
            class="MuiPrivateTextarea-textarea-144 MuiPrivateTextarea-shadow-145"
            readonly=""
            rows="1"
            tabindex="-1"
          />
          <textarea
            aria-hidden="true"
            class="MuiPrivateTextarea-textarea-144 MuiPrivateTextarea-shadow-145"
            readonly=""
            rows="1"
            tabindex="-1"
          />
          <textarea
            aria-invalid="false"
            class="MuiPrivateTextarea-textarea-144 MuiInputBase-input-134 MuiOutlinedInput-input-119 MuiInputBase-inputMultiline-136 MuiOutlinedInput-inputMultiline-121"
            rows="1"
            style="height:19px"
          />
        </div>
      </div>
    </div>
    <button
      class="MuiButtonBase-root-28 MuiButton-root-2 MuiButton-contained-13 MuiButton-raised-16"
      style="width:50%;margin-top:15px"
      tabindex="0"
      type="submit"
    >
      <span
        class="MuiButton-label-3"
      >
        Submit
      </span>
    </button>
  </div>
</div>
`;
